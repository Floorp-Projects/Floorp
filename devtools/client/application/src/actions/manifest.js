/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const { l10n } = require("devtools/client/application/src/modules/l10n");

const {
  services,
  ManifestDevToolsError,
} = require("devtools/client/application/src/modules/application-services");
const {
  FETCH_MANIFEST_FAILURE,
  FETCH_MANIFEST_START,
  FETCH_MANIFEST_SUCCESS,
  RESET_MANIFEST,
} = require("devtools/client/application/src/constants");

function fetchManifest() {
  return async ({ dispatch, getState }) => {
    dispatch({ type: FETCH_MANIFEST_START });
    try {
      const manifest = await services.fetchManifest();
      dispatch({ type: FETCH_MANIFEST_SUCCESS, manifest });
    } catch (error) {
      let errorMessage = error.message;

      // since Firefox DevTools errors may not make sense for the user, swap
      // their message for a generic one.
      if (error instanceof ManifestDevToolsError) {
        console.error(error);
        errorMessage = l10n.getString("manifest-loaded-devtools-error");
      }

      dispatch({ type: FETCH_MANIFEST_FAILURE, error: errorMessage });
    }
  };
}

function resetManifest() {
  return { type: RESET_MANIFEST };
}

module.exports = {
  fetchManifest,
  resetManifest,
};
