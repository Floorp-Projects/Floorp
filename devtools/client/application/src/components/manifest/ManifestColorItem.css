/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.manifest-item__color {
  /* NOTE: platform converts any color format that is in the manifest to
           hexadecimal, so we can uppercase */
  text-transform: uppercase;
  direction: ltr; /* force LTR so the # stays at the beginning of the hex number */
  display: inline-block;
}

.manifest-item__color::before {
  display: inline-block;
  content: '';
  background-color: #fff;
  background-image: linear-gradient(var(--color-value), var(--color-value)), /* injected via React */
    linear-gradient(45deg, #ccc 25%, transparent 25%, transparent 75%, #ccc 75%, #ccc),
    linear-gradient(45deg, #ccc 25%, transparent 25%, transparent 75%, #ccc 75%, #ccc);
  background-size: 6px 6px;
  background-position: 0 0, 3px 3px;
  border-radius: 50%;
  width: calc(var(--base-unit) * 3);
  height: calc(var(--base-unit) * 3);
  margin-block-start: calc(var(--base-unit) * -0.5);
  margin-inline-end: var(--base-unit);
  box-shadow: 0 0 0 1px var(--theme-splitter-color);
  vertical-align: middle;
}
