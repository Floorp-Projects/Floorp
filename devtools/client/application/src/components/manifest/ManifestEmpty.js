/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const { openDocLink } = require("devtools/client/shared/link");

const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const {
  a,
  article,
  aside,
  div,
  h1,
  img,
  p,
} = require("devtools/client/shared/vendor/react-dom-factories");

const FluentReact = require("devtools/client/shared/vendor/fluent-react");
const Localized = createFactory(FluentReact.Localized);

const DOC_URL =
  "https://developer.mozilla.org/en-US/docs/Web/Manifest" +
  "?utm_source=devtools&utm_medium=sw-panel-blank";

/**
 * This component displays help information when no manifest is found for the
 * current target.
 */
class ManifestEmpty extends PureComponent {
  openDocumentation() {
    openDocLink(DOC_URL);
  }

  render() {
    return article(
      { className: "app-page__icon-container js-manifest-empty" },
      aside(
        {},
        Localized(
          {
            id: "sidebar-item-manifest",
            attrs: {
              alt: true,
            },
          },
          img({
            className: "app-page__icon",
            src: "chrome://devtools/skin/images/application-manifest.svg",
          })
        )
      ),
      div(
        {},
        Localized(
          {
            id: "manifest-empty-intro2",
          },
          h1({ className: "app-page__title" })
        ),
        p(
          {},
          Localized(
            { id: "manifest-empty-intro-link" },
            a({
              onClick: () => this.openDocumentation(),
            })
          )
        ),
        Localized({ id: "manifest-non-existing" }, p({}))
      )
    );
  }
}

// Exports
module.exports = ManifestEmpty;
