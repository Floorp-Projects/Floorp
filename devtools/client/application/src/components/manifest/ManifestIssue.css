/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.manifest-issue__icon {
  -moz-context-properties: fill;
  width: calc(var(--base-unit) * 3);
  height: calc(var(--base-unit) * 3);
}

.manifest-issue__icon--warning {
  fill: var(--theme-icon-warning-color);
}

.manifest-issue__icon--error {
  fill: var(--theme-icon-error-color);
}
