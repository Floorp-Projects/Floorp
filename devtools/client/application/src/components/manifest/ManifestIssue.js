/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const {
  img,
  li,
  span,
} = require("devtools/client/shared/vendor/react-dom-factories");

const FluentReact = require("devtools/client/shared/vendor/fluent-react");
const Localized = createFactory(FluentReact.Localized);

const {
  MANIFEST_ISSUE_LEVELS,
} = require("devtools/client/application/src/constants");
const Types = require("devtools/client/application/src/types/index");

/**
 * A Manifest validation issue (warning, error)
 */
class ManifestIssue extends PureComponent {
  static get propTypes() {
    return {
      className: PropTypes.string,
      ...Types.manifestIssue, // { level, message }
    };
  }

  getIconData(level) {
    switch (level) {
      case MANIFEST_ISSUE_LEVELS.WARNING:
        return {
          src: "chrome://devtools/skin/images/alert-small.svg",
          localizationId: "icon-warning",
        };
      case MANIFEST_ISSUE_LEVELS.ERROR:
      default:
        return {
          src: "chrome://devtools/skin/images/error-small.svg",
          localizationId: "icon-error",
        };
    }
  }

  render() {
    const { level, message, className } = this.props;
    const icon = this.getIconData(level);

    return li(
      { className: `js-manifest-issue ${className ? className : ""}` },
      Localized(
        { id: icon.localizationId, attrs: { alt: true, title: true } },
        img({
          className: `manifest-issue__icon manifest-issue__icon--${level}`,
          src: icon.src,
        })
      ),
      span({}, message)
    );
  }
}

// Exports
module.exports = ManifestIssue;
