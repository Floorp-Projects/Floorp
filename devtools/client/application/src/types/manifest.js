/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const PropTypes = require("devtools/client/shared/vendor/react-prop-types");

const {
  MANIFEST_ISSUE_LEVELS,
} = require("devtools/client/application/src/constants");
const {
  MANIFEST_MEMBER_VALUE_TYPES,
} = require("devtools/client/application/src/constants");

const manifestIssue = {
  level: PropTypes.oneOf(Object.values(MANIFEST_ISSUE_LEVELS)).isRequired,
  message: PropTypes.string.isRequired,
  // NOTE: we are currently ignoring the 'type' field that platform adds to errors
};

const manifestIssueArray = PropTypes.arrayOf(PropTypes.shape(manifestIssue));

const manifestItemColor = {
  label: PropTypes.string.isRequired,
  value: PropTypes.string,
};

const manifestItemIcon = {
  label: PropTypes.shape({
    contentType: PropTypes.string,
    sizes: PropTypes.string,
  }).isRequired,
  value: PropTypes.shape({
    src: PropTypes.string.isRequired,
    purpose: PropTypes.string.isRequired,
  }).isRequired,
};

const manifestItemUrl = {
  label: PropTypes.string.isRequired,
  value: PropTypes.string,
};

const manifestMemberColor = {
  key: manifestItemColor.label,
  value: manifestItemColor.value,
  type: PropTypes.oneOf([MANIFEST_MEMBER_VALUE_TYPES.COLOR]),
};

const manifestMemberIcon = {
  key: manifestItemIcon.label,
  value: manifestItemIcon.value,
  type: PropTypes.oneOf([MANIFEST_MEMBER_VALUE_TYPES.ICON]),
};

const manifestMemberString = {
  key: PropTypes.string.isRequired,
  value: PropTypes.string,
  type: PropTypes.oneOf([MANIFEST_MEMBER_VALUE_TYPES.STRING]),
};

const manifest = {
  // members
  identity: PropTypes.arrayOf(PropTypes.shape(manifestMemberString)).isRequired,
  presentation: PropTypes.arrayOf(
    PropTypes.oneOfType([
      PropTypes.shape(manifestMemberColor),
      PropTypes.shape(manifestMemberString),
    ])
  ).isRequired,
  icons: PropTypes.arrayOf(PropTypes.shape(manifestMemberIcon)).isRequired,
  // validation issues
  validation: manifestIssueArray.isRequired,
  // misc
  url: PropTypes.string.isRequired,
};

module.exports = {
  // full manifest
  manifest,
  // specific manifest items
  manifestItemColor,
  manifestItemIcon,
  manifestItemUrl,
  // manifest issues
  manifestIssue,
  manifestIssueArray,
};
