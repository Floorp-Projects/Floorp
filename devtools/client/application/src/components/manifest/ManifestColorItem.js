/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const { div } = require("devtools/client/shared/vendor/react-dom-factories");

const Types = require("devtools/client/application/src/types/index");
const ManifestItem = createFactory(
  require("devtools/client/application/src/components/manifest/ManifestItem")
);

/**
 * This component displays a Manifest member which holds a color value
 */
class ManifestColorItem extends PureComponent {
  static get propTypes() {
    return {
      ...Types.manifestItemColor, // { label, value }
    };
  }

  renderColor() {
    let { value } = this.props;
    if (!value) {
      return null;
    }

    // Truncate colors in #rrggbbaa format to #rrggbb
    if (value.length === 9 && value.toLowerCase().endsWith("ff")) {
      value = value.slice(0, 7);
    }

    /* div instead of span because CSS `direction` works with block elements */
    return div(
      {
        className: "manifest-item__color",
        style: { "--color-value": value },
      },
      value
    );
  }

  render() {
    const { label } = this.props;
    return ManifestItem({ label }, this.renderColor());
  }
}

module.exports = ManifestColorItem;
