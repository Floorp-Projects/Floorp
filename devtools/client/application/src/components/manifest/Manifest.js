/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const {
  article,
  h1,
  table,
  tbody,
} = require("devtools/client/shared/vendor/react-dom-factories");

const FluentReact = require("devtools/client/shared/vendor/fluent-react");
const Localized = createFactory(FluentReact.Localized);
const { l10n } = require("devtools/client/application/src/modules/l10n");

const ManifestColorItem = createFactory(
  require("devtools/client/application/src/components/manifest/ManifestColorItem")
);
const ManifestIconItem = createFactory(
  require("devtools/client/application/src/components/manifest/ManifestIconItem")
);
const ManifestUrlItem = createFactory(
  require("devtools/client/application/src/components/manifest/ManifestUrlItem")
);
const ManifestItem = createFactory(
  require("devtools/client/application/src/components/manifest/ManifestItem")
);
const ManifestIssueList = createFactory(
  require("devtools/client/application/src/components/manifest/ManifestIssueList")
);
const ManifestSection = createFactory(
  require("devtools/client/application/src/components/manifest/ManifestSection")
);
const ManifestJsonLink = createFactory(
  require("devtools/client/application/src/components/manifest/ManifestJsonLink")
);

const {
  MANIFEST_MEMBER_VALUE_TYPES,
} = require("devtools/client/application/src/constants");
const Types = require("devtools/client/application/src/types/index");

/**
 * A canonical manifest, splitted in different sections
 */
class Manifest extends PureComponent {
  static get propTypes() {
    return {
      ...Types.manifest, // { identity, presentation, icons, validation, url }
    };
  }

  renderIssueSection() {
    const { validation } = this.props;
    const shouldRender = validation && validation.length > 0;

    return shouldRender
      ? ManifestSection(
          {
            key: `manifest-section-0`,
            title: l10n.getString("manifest-item-warnings"),
          },
          ManifestIssueList({ issues: validation })
        )
      : null;
  }

  renderMember({ key, value, type }, index) {
    let domKey = key;
    switch (type) {
      case MANIFEST_MEMBER_VALUE_TYPES.COLOR:
        return ManifestColorItem({ label: key, key: domKey, value });
      case MANIFEST_MEMBER_VALUE_TYPES.ICON:
        // since the manifest may have keys with empty size/contentType,
        // we cannot use them as unique IDs
        domKey = index;
        return ManifestIconItem({ label: key, key: domKey, value });
      case MANIFEST_MEMBER_VALUE_TYPES.URL:
        return ManifestUrlItem({ label: key, key: domKey, value });
      case MANIFEST_MEMBER_VALUE_TYPES.STRING:
      default:
        return ManifestItem({ label: key, key: domKey }, value);
    }
  }

  renderItemSections() {
    const { identity, icons, presentation } = this.props;

    const manifestMembers = [
      { localizationId: "manifest-item-identity", items: identity },
      { localizationId: "manifest-item-presentation", items: presentation },
      { localizationId: "manifest-item-icons", items: icons },
    ];

    return manifestMembers.map(({ localizationId, items }, index) => {
      return ManifestSection(
        {
          key: `manifest-section-${index + 1}`,
          title: l10n.getString(localizationId),
        },
        // NOTE: this table should probably be its own component, to keep
        //       the same level of abstraction as with the validation issues
        // Bug https://bugzilla.mozilla.org/show_bug.cgi?id=1577138
        table({}, tbody({}, items.map(this.renderMember)))
      );
    });
  }

  render() {
    const { url } = this.props;

    return article(
      { className: "js-manifest" },
      Localized(
        {
          id: "manifest-view-header",
        },
        h1({ className: "app-page__title" })
      ),
      ManifestJsonLink({ url }),
      this.renderIssueSection(),
      this.renderItemSections()
    );
  }
}

// Exports
module.exports = Manifest;
