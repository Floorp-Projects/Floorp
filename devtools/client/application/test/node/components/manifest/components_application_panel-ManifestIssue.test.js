/* Any copyright is dedicated to the Public Domain.
   http://creativecommons.org/publicdomain/zero/1.0/ */

"use strict";

// Import libs
const { shallow } = require("enzyme");
const { createFactory } = require("react");

const ManifestIssue = createFactory(
  require("devtools/client/application/src/components/manifest/ManifestIssue")
);

/*
 * Tests for the ManifestIssue component
 */

describe("ManifestIssue", () => {
  it("renders the expected snapshot for a warning", () => {
    const issue = { level: "warning", message: "Lorem ipsum" };
    const wrapper = shallow(ManifestIssue(issue));
    expect(wrapper).toMatchSnapshot();
  });

  it("renders the expected snapshot for an error", () => {
    const issue = { level: "error", message: "Lorem ipsum" };
    const wrapper = shallow(ManifestIssue(issue));
    expect(wrapper).toMatchSnapshot();
  });
});
