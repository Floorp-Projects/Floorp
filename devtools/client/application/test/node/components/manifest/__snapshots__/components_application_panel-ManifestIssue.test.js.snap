// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManifestIssue renders the expected snapshot for a warning 1`] = `
<li
  className="js-manifest-issue "
>
  <Localized
    attrs={
      Object {
        "alt": true,
        "title": true,
      }
    }
    id="icon-warning"
  >
    <img
      className="manifest-issue__icon manifest-issue__icon--warning"
      src="chrome://devtools/skin/images/alert-small.svg"
    />
  </Localized>
  <span>
    Lorem ipsum
  </span>
</li>
`;

exports[`ManifestIssue renders the expected snapshot for an error 1`] = `
<li
  className="js-manifest-issue "
>
  <Localized
    attrs={
      Object {
        "alt": true,
        "title": true,
      }
    }
    id="icon-error"
  >
    <img
      className="manifest-issue__icon manifest-issue__icon--error"
      src="chrome://devtools/skin/images/error-small.svg"
    />
  </Localized>
  <span>
    Lorem ipsum
  </span>
</li>
`;
